libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefresherExecutor.kt

```diff
@@ -0,0 +1,41 @@
+package com.mapbox.navigation.core.routerefresh
+
+import com.mapbox.navigation.base.route.NavigationRoute
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+internal interface RouteRefresherProgressCallback {
+
+    fun onStarted()
+
+    fun onResult(routeRefresherResult: RouteRefresherResult)
+}
+
+internal class RouteRefresherExecutor(
+    private val routeRefresher: RouteRefresher,
+    private val scope: CoroutineScope,
+    private val timeout: Long,
+) {
+
+    private val mutex = Mutex()
+    private val queue = ArrayDeque<Pair<List<NavigationRoute>, RouteRefresherProgressCallback>>()
```

Comment:
Why do we need a queue here? What type of requests overlap and need to be scheduled this way?

Reply:
That's what we discussed: if we invoke `refreshRoutesImmediately` very often, we should:
1) have only 1 running request at a time;
2) execute only current request (if present) and the latest one, not all of them.

Reply:
`refreshRoutesImmediately` has no callback in the current proposal, it's a call that developers make and forget, the call provides no direct side effects (there's no data that developers provide to us and expect us to apply where a coalescing pattern would be needed), and they have no means to track the state and result of it anyway (they don't need an exact ratio of call to callback). This makes me think that we don't need this complexity - if there's already a refresh request in the air when `refreshRoutesImmediately` is called, we can ignore the new call.

Reply:
OK, makes sense.